sdk/digitaltwins/azure-digitaltwins-core/src/main/java/com/azure/digitaltwins/core/DigitalTwinsAsyncClient.java

```diff
@@ -1067,4 +1068,142 @@ public HttpPipeline getHttpPipeline() {
+            nextLink -> withContext(context -> queryNextPage(query, nextLink, clazz, context)));
+    }
+
+    <T> PagedFlux<T> query(String query, Class<T> clazz, Context context) {
+        return new PagedFlux<>(
+            () -> queryFirstPage(query, clazz, context),
+            nextLink -> queryNextPage(query, nextLink, clazz, context));
+    }
+
+    Mono<PagedResponse<String>> queryFirstPage(String query, Context context) {
+        QuerySpecification querySpecification = new QuerySpecification();
+
+        querySpecification
+            .setQuery(query);
+
+        return protocolLayer
+            .getQueries()
+            .queryTwinsWithResponseAsync(querySpecification, context)
+            .map(objectPagedResponse -> new PagedResponseBase<>(
+                objectPagedResponse.getRequest(),
+                objectPagedResponse.getStatusCode(),
+                objectPagedResponse.getHeaders(),
+                objectPagedResponse.getValue().getItems().stream()
+                    .map(object -> {
+                        try {
+                            return mapper.writeValueAsString(object);
+                        } catch (JsonProcessingException e) {
+                            logger.error("JsonProcessingException occurred while retrieving query result items: ", e);
+                            throw new RuntimeException("JsonProcessingException occurred while retrieving query result items", e);
+                        }
+                    })
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toList()),
+                objectPagedResponse.getValue().getContinuationToken(),
+                objectPagedResponse.getDeserializedHeaders()));
+    }
+
+    <T> Mono<PagedResponse<T>> queryFirstPage(String query, Class<T> clazz, Context context) {
+        QuerySpecification querySpecification = new QuerySpecification();
+
+        querySpecification
+            .setQuery(query);
+
+        return protocolLayer
+            .getQueries()
+            .queryTwinsWithResponseAsync(querySpecification, context)
+            .map(objectPagedResponse -> new PagedResponseBase<>(
+                objectPagedResponse.getRequest(),
+                objectPagedResponse.getStatusCode(),
+                objectPagedResponse.getHeaders(),
+                objectPagedResponse.getValue().getItems().stream()
+                    .map(object -> mapper.convertValue(object, clazz))
+                    .filter(Objects::nonNull)
+                    .collect(Collectors.toList()),
+                objectPagedResponse.getValue().getContinuationToken(),
+                objectPagedResponse.getDeserializedHeaders()));
+    }
+
+    Mono<PagedResponse<String>> queryNextPage(String query, String nextLink, Context context) {
+        QuerySpecification querySpecification = new QuerySpecification();
+
+        querySpecification
+            .setQuery(query)
+            .setContinuationToken(nextLink);
```

Comment:
From the .NET client, it looks like all we need to supply to the service is the cont. token, we don't need to pass in the query string again: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/digitaltwins/Azure.DigitalTwins.Core/src/DigitalTwinsClient.cs#L1360-L1363

Also, what is our testing strategy with this (the pageable part)? Do we create 100+ twin instances and then query them? The page size isn't configurable yet, correct?

Reply:
yes that is what I replied to your earlier comment. You don't need to pass it but there is no harm in passing it. But to keep consistency I removed it here as well. The page size is not configurable but we will follow the same strategy as we have in .net. I will look at it as I start implementing samples.